src/Riok.Mapperly/Descriptors/MappingBuilders/EnumToEnumMappingBuilder.cs

```diff
@@ -283,44 +190,69 @@ private static EnumFallbackValueMapping BuildFallbackMapping(MappingBuilderConte
 
     private static IReadOnlyDictionary<IFieldSymbol, IFieldSymbol> BuildExplicitValueMappings(MappingBuilderContext ctx)
     {
-        var targetFieldsByExplicitValue = new Dictionary<IFieldSymbol, IFieldSymbol>(SymbolEqualityComparer.Default);
+        var explicitMappings = new Dictionary<IFieldSymbol, IFieldSymbol>(SymbolEqualityComparer.Default);
         foreach (var (source, target) in ctx.Configuration.Enum.ExplicitMappings)
         {
-            if (!SymbolEqualityComparer.Default.Equals(source.Type, ctx.Source))
+            if (source.ConstantValue.Kind is not TypedConstantKind.Enum)
+            {
+                ctx.ReportDiagnostic(
+                    DiagnosticDescriptors.MapValueTypeMismatch,
+                    source.Expression.ToFullString(),
+                    source.ConstantValue.Type?.ToDisplayString() ?? "unknown",
+                    ctx.Source
+                );
+                continue;
+            }
+
+            if (target.ConstantValue.Kind is not TypedConstantKind.Enum)
+            {
+                ctx.ReportDiagnostic(
+                    DiagnosticDescriptors.MapValueTypeMismatch,
+                    target.Expression.ToFullString(),
+                    target.ConstantValue.Type?.ToDisplayString() ?? "unknown",
+                    ctx.Target
+                );
+                continue;
+            }
+
+            var sourceField = source.ConstantValue.GetEnumField()!;
+            var targetField = target.ConstantValue.GetEnumField()!;
+
+            if (!SymbolEqualityComparer.Default.Equals(source.ConstantValue.Type, ctx.Source))
             {
                 ctx.ReportDiagnostic(
                     DiagnosticDescriptors.SourceEnumValueDoesNotMatchSourceEnumType,
-                    source,
-                    source.ConstantValue ?? 0,
-                    source.Type,
+                    sourceField,
+                    source.ConstantValue.Value ?? 0,
+                    target.ConstantValue.Type?.ToDisplayString() ?? "unknown",
                     ctx.Source
                 );
                 continue;
             }
 
-            if (!SymbolEqualityComparer.Default.Equals(target.Type, ctx.Target))
+            if (!SymbolEqualityComparer.Default.Equals(target.ConstantValue.Type, ctx.Target))
             {
                 ctx.ReportDiagnostic(
                     DiagnosticDescriptors.TargetEnumValueDoesNotMatchTargetEnumType,
-                    target,
-                    target.ConstantValue ?? 0,
-                    target.Type,
+                    targetField,
+                    target.ConstantValue.Value ?? 0,
+                    target.ConstantValue.Type?.ToDisplayString() ?? "unknown",
                     ctx.Target
                 );
                 continue;
             }
 
-            if (targetFieldsByExplicitValue.ContainsKey(source))
+            if (explicitMappings.ContainsKey(sourceField))
```

Comment:
Nitpick: I just noted there is the dictionary extension method `TryAdd` which does exactly this (`ContainsKey` + `Add`), no else would be needed then.

Reply:
Yes, agreed. I have made this change.

Reply:
This is not fixed, is it? If condition should be `!explicitMappings.TryAdd(sourceField, targetField)` with no else body.